core/src/main/java/com/linecorp/armeria/internal/common/SplitHttpMessageSubscriber.java

```diff
@@ -0,0 +1,301 @@
+            return trailersFuture;
+        }
+
+        trailersFuture = new HeadersFuture<>();
+        if (trailersFutureUpdater.compareAndSet(this, null, trailersFuture)) {
+            return trailersFuture;
+        } else {
+            return this.trailersFuture;
+        }
+    }
+
+    final boolean wroteAny() {
+        return wroteAny;
+    }
+
+    @Nullable
+    final Subscription upstream() {
+        return upstream;
+    }
+
+    void initDownstream(Subscriber<? super HttpData> downstream, EventExecutor downstreamExecutor,
+                        SubscriptionOption... options) {
+        assert downstreamExecutor.inEventLoop();
+
+        this.downstream = downstream;
+        this.downstreamExecutor = downstreamExecutor;
+        for (SubscriptionOption option : options) {
+            if (option == SubscriptionOption.NOTIFY_CANCELLATION) {
+                notifyCancellation = true;
+            } else if (option == SubscriptionOption.WITH_POOLED_OBJECTS) {
+                usePooledObject = true;
+            }
+        }
+
+        try {
+            downstream.onSubscribe(this);
+            final Throwable cause = this.cause;
+            if (cause != null) {
+                onError0(cause, downstream);
+            } else if (completing) {
+                downstream.onComplete();
+            }
+        } catch (Throwable t) {
+            throwIfFatal(t);
+            logger.warn("Subscriber should not throw an exception. subscriber: {}", downstream, t);
+        }
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (upstream != null) {
+            subscription.cancel();
+            return;
+        }
+        upstream = subscription;
+        if (cancelCalled) {
+            subscription.cancel();
+            return;
+        }
+        if (pendingRequests > 0) {
+            subscription.request(pendingRequests);
+        }
+    }
+
+    @Override
+    public void request(long n) {
+        if (n <= 0) {
+            upstreamMessage.abort(new IllegalArgumentException(
+                    "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+            return;
+        }
+        if (upstreamExecutor.inEventLoop()) {
+            request0(n);
+        } else {
+            upstreamExecutor.execute(() -> request0(n));
+        }
+    }
+
+    private void request0(long n) {
+        final Subscription upstream = this.upstream;
+        if (upstream == null) {
+            pendingRequests = LongMath.saturatedAdd(n, pendingRequests);
+        } else {
+            upstream.request(n);
+        }
+    }
+
+    @Override
+    public final void cancel() {
+        if (cancelCalled) {
+            return;
+        }
+        cancelCalled = true;
+        if (!notifyCancellation) {
+            downstream = NoopSubscriber.get();
+        }
+        completeTrailers(HttpHeaders.of());
```

Comment:
Q) Is there a particular reason we use `completeTrailers(HttpHeaders.of())` instead of `doOnCompletion(null);`?

The implementation will be slightly different from the previous approach where:
1) trailers will be completed immediately
2) headers will be completed after `upstream.cancel()` -> `onError`

I think this has a slight nuance where `trailers` can be completed before `headers` depending on how the upstream is implemented which I find surprising. (the following test fails)

```
    @Test
    void completeHeadersBeforeTrailer() {
        final HttpResponse response = HttpResponse.delayed(HttpResponse.of(HttpStatus.OK), Duration.ofSeconds(1));
        final SplitHttpResponse bodyStream = response.split();
        bodyStream.body().subscribe(new Subscriber<HttpData>() {
            @Override
            public void onSubscribe(Subscription s) {
                s.cancel();
            }

            @Override
            public void onNext(HttpData httpData) {}

            @Override
            public void onError(Throwable t) {}

            @Override
            public void onComplete() {}
        }, new DefaultEventLoop());
        final AtomicBoolean atomicBoolean = new AtomicBoolean();
        bodyStream.trailers().whenComplete((h, throwable) -> {
            atomicBoolean.set(bodyStream.headers().isDone());
        }).join();
        assertThat(atomicBoolean).isTrue();
    }
```

Reply:
Calling `completeTrailers(HttpHeaders.of())` in the `cancel()` method is probably a mistake. I'm sorry :-(

According to the current implementation, the trailer can be completed before the header when `subscription.cancel()` is called, it can be an issue, right?

I thought for a moment, can't it be resolved by not calling both `completeTrailers(HttpHeaders.of())` and `doOnCompletion(null)` in the `cancel()` method...? This is because if methods are called in the order of `upstream.cancel()` -> `onError(cause)`, the `onError(cause)` completes the header and trailer.

Then it is unlikely that the trailer will complete before the header.

Reply:
> How about not calling both completeTrailers(HttpHeaders.of()) and doOnCompletion(null) in the cancel() method? This is because if methods are called in the order of upstream.cancel() -> onError(cause), the onError(cause) completes the header and trailer.

`+1` for this. Could you also add a comment on why we don't call `doOnCompletion` explicitly?

Reply:
@minwoox, As you said, it might be a good idea to explicitly call `doOnCompletion` to safely complete the header.
I fixed it in this commit. https://github.com/line/armeria/pull/3953/commits/c0e2a1ccaaaccfad07a092298a70561407bcd095

This is just an idea, how about not explicitly calling `doOnCompletion` in the `cancel()` method like below?
```java
class SplitHttpMessageSubscriber implements Subscriber<HttpObject>, Subscription {
    ...
    @Override
    public final void cancel() {
        if (cancelCalled) {
            return;
        }
        cancelCalled = true;
        if (!notifyCancellation) {
            downstream = NoopSubscriber.get();
        }
        // doOnCompletion(null); <-- Like this, how about not to call this explicitly?
        final Subscription upstream = this.upstream;
        if (upstream != null) {
            upstream.cancel(); // After this line, onError(cause) calls doOnCompletion(cause) explicitly.
        }
    }
}
```
Because when called `upstream.cancel()` -> `onError(cause)`, `onError(cause)` explicitly calls `doOnCompletion(cause)`. And this time the cause is `CancelledSubscriptionException`, so `headersFuture.doComplete()` is called.

Therefore, explicitly calling `doOnCompletion(null)` in the `cancel()` method and not calling `doOnCompletion(null)` is expected to have the same result.

Reply:
> Therefore, explicitly calling doOnCompletion(null) in the cancel() method and not calling doOnCompletion(null) is expected to have the same result.

Yeah, that's exactly what I meant. 😅 
Let's not call `doOnCompletion` and handle it in `onError(cause)`. 
It might be helpful for reviewers if we leave a comment: `// doOnCompletion is called in onError(cause) eventually.`

Reply:
Ah, I misunderstood your comment, sorry :(
I fixed it by not calling `doOnCompletion(null)` directly in [this commit](https://github.com/line/armeria/pull/3953/commits/ada15ad1035584250986910928d163dd5739bdb2). Thanks for the good comments! 👍